### PR TITLE
Adding fzf ebuild.

### PR DESCRIPTION
**Pull request by rafspiny**
Adding ebuild for a new package.

Gentoo Bug: [703082](https://bugs.gentoo.org/703082)
There is one warning that I tried to address, but was not able to. More info on this post.
https://forums.gentoo.org/viewtopic-t-1110938-highlight-.html

### CODE REVIEW

File: app-shells/fzf/fzf-0.21.0.ebuild [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+)
+go-module_set_globals
+
+EGO_PN="github.com/junegunn/${PN}"
+SRC_URI="https://${EGO_PN}/archive/${PV}.tar.gz -> ${P}.tar.gz
+			${EGO_SUM_SRC_URI}"
+
+LICENSE="MIT"
+SLOT="0"
+KEYWORDS="~amd64 ~arm ~x86"
```
Comment (juippis) [inline app-shells/fzf/fzf-0.21.0.ebuild:65]: You can only add ~amd64 and x86, provided you tested your ebuild with both arches. Rest **need** to be keyworded through respected arch teams.

File: app-shells/fzf/fzf-0.21.0.ebuild [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+IUSE="fish-completion neovim tmux vim zsh-completion"
```
Comment (juippis) [inline app-shells/fzf/fzf-0.21.0.ebuild:66]: Add these as `optfeature` and always installed their completion files.
https://projects.gentoo.org/qa/policy-guide/installed-files.html#pg0301

File: app-shells/fzf/fzf-0.21.0.ebuild [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+DEPEND=""
+RDEPEND="
+	${DEPEND}
+	fish-completion? ( app-shells/fish )
+	tmux? ( app-misc/tmux )
+	neovim? ( app-editors/neovim )
+	vim? ( app-editors/vim )
+	zsh-completion? ( app-shells/zsh )
+"
+BDEPEND=""
```
Comment (juippis) [inline app-shells/fzf/fzf-0.21.0.ebuild:77]: All can most likely be removed. Use `optfeature` instead.

File: app-shells/fzf/fzf-0.21.0.ebuild [hunk 4]
```diff
@@ -0,0 +1,124 @@
+# Copyright 2019-2020 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
+
+inherit go-module bash-completion-r1
```
Comment (juippis) [inline app-shells/fzf/fzf-0.21.0.ebuild:6]: You can use an eclass for `zsh` too.

File: app-shells/fzf/fzf-0.21.0.ebuild [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+		insinto /usr/share/zsh/site-functions
+		newins shell/completion.zsh _${PN}
+		insinto /usr/share/zsh/site-contrib/
+		newins shell/key-bindings.zsh ${PN}.zsh
+	fi
+
+	# Install bash completion file and key-bindings
+	newbashcomp shell/completion.bash ${PN}
+#	mkdir -p "${ED}$(get_bashcompdir)/"
+#	cp shell/completion.bash "${ED}$(get_bashcompdir)/${PN}"
```
Comment (juippis) [inline app-shells/fzf/fzf-0.21.0.ebuild:121]: Debugging lines should be removed when it works.

File: app-shells/fzf/metadata.xml [hunk 6]
```diff
@@ -0,0 +1,39 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE pkgmetadata SYSTEM "http://www.gentoo.org/dtd/metadata.dtd">
+<pkgmetadata>
+	<maintainer type="person">
+		<email>mgorny@gentoo.org</email>
+		<name>Michał Górny</name>
+	</maintainer>
+	<maintainer type="project">
+		<email>shell-tools@gentoo.org</email>
+		<name>Gentoo Shell Tools Project</name>
```
Comment (juippis) [inline app-shells/fzf/metadata.xml:10]: Are you just adding random people in here hoping they'd start to maintain it? 
You need to become the maintainer unless you have their consent on being added here. You can commit and maintain this package through proxy-maint project, 
https://wiki.gentoo.org/wiki/Project:Proxy_Maintainers/User_Guide#Proxied_maintainer_in_metadata.xml

**Comment by juippis** [comment]
I also failed to build it, 
```
# github.com/junegunn/fzf
flag provided but not defined: -Wl,--as-needed
usage: link [options] main.o
make: *** [Makefile:103: target/fzf-linux_amd64] Error 2
 * ERROR: app-shells/fzf-0.21.0::testworld failed (compile phase):
 *   emake failed
```

Then you need to read https://www.gentoo.org/glep/glep-0066.html about good commit summaries, and https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin which is required for any commit to be merged into ::gentoo.

File: app-shells/fzf/fzf-0.21.0.ebuild [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+	"golang.org/x/tools v0.0.0-20191011211836-4c025a95b26e/go.mod h1:b+2E5dAYhXwXZwtnZ6UAqBI28+e2cm9otk0dWdXHAEo="
+	"golang.org/x/xerrors v0.0.0-20190717185122-a985d3407aa7/go.mod h1:I/5z698sn9Ka8TeJc9MKroUUfqBBauWjQqLJ2OPfmY0="
+)
+go-module_set_globals
+
+EGO_PN="github.com/junegunn/${PN}"
+SRC_URI="https://${EGO_PN}/archive/${PV}.tar.gz -> ${P}.tar.gz
+			${EGO_SUM_SRC_URI}"
+
+LICENSE="MIT"
```
Comment (juippis) [inline app-shells/fzf/fzf-0.21.0.ebuild:63]: And one more thing: You need to include licenses for ALL bundled modules. I think there's an ebuild to check these, go-license-detector or something.

File: app-shells/fzf/fzf-0.21.0.ebuild (hunk 2 above)
Comment (rafspiny) [reply app-shells/fzf/fzf-0.21.0.ebuild:66]: @juippis do you refer to `*-completion` or all of them?
They all install small files. So I guess the second option.

File: app-shells/fzf/metadata.xml (hunk 6 above)
Comment (rafspiny) [reply app-shells/fzf/metadata.xml:10]: Copy and paste error

File: app-shells/fzf/fzf-0.21.0.ebuild (hunk 4 above)
Comment (rafspiny) [reply app-shells/fzf/fzf-0.21.0.ebuild:6]: I can't fins zsh eclass listed here: 
https://devmanual.gentoo.org/eclass-reference/

File: app-shells/fzf/fzf-0.21.0.ebuild (hunk 5 above)
Comment (rafspiny) [reply app-shells/fzf/fzf-0.21.0.ebuild:121]: Removed

**Comment by rafspiny** [comment]
umh. That compilation error is new to me. 
I will double-check LDFLAGS

**Comment by Alessandro-Barbieri** [comment]
See also #15407

**Comment by heroin-moose** [comment]
@rafspiny, the compilation error is due to missing quotes around -extldflags. See https://github.com/junegunn/fzf/pull/1991.

**Comment by rafspiny** [comment]
I had time to look into it today. Fixed it and pushed.
@heroin-moose I arrived to the same conclusion. I created an issue on fzf too and submitted a PR. I looked for LDFLAGS in their issues, but not popped up : /
Oh well, it was not a lot of work in the end.

Everything shall be good now.

**Comment by juippis** [comment]
Actually nope, you need to sign-off your commits. And please squash them into one. Describe the package in few words in git message box.

**Comment by rafspiny** [comment]
@juippis Sure thing. I will do it tonight

**Comment by heroin-moose** [comment]
Is setting S is required? I believe not.

**Comment by heroin-moose** [comment]
@juippis, also fzf-tmux will fail if tmux is not installed. Perhaps it makes sense to depend on tmux?

**Comment by rafspiny** [comment]
@juippis I rebased and added the `Signed-off-by` tag.
@heroin-moose yep, that secondary binary `fzf-tmux` will fail without tmux. The main binary, `fzf` will still work though.
Do you suggest to put **tmux** as a RDEPEND use flag?

**Comment by heroin-moose** [comment]
Yeah, like this:

```
RDEPEND="tmux? ( app-misc/tmux )"
```

**Comment by rafspiny** [comment]
@heroin-moose yep. The if part was in my initial commit when `tmux` was still in the `DEPEND` variable.

I am not sure if we shall use the USE flag or not. In the end, that script is supposed to be called from the shell. It is not integrated into other commands. 
On the other hand, I very much dislike having a script on my system that I know is broken.
Probably, the best thing is to re-add the `tmux` use flag.

@juippis what's your suggestions here?

**Comment by heroin-moose** [comment]
Yeah, I also don't the like idea of broken scripts in /usr/bin. Along with the man page that states "hey, run this thing, it's good".

**Comment by heroin-moose** [comment]
Well... it really works and ignores so maybe there is really no need in USE flag after all.

**Comment by rafspiny** [comment]
It does indeed. Even better!

**Comment by heroin-moose** [comment]
One more thing: should we install keybindings to /etc for bas? It will overwrite the default ones, which I'm sure that most people want. Maybe it would be better to put all keybindings to /usr/share/fzf.

**Comment by rafspiny** [comment]
@heroin-moose I am not sure about that. 
I am doing:
> insinto /etc/bash/bashrc.d/
> newins shell/key-bindings.bash ${PN}.sh

That shall install the file in /etc/bash/bashrc.d/fzf.sh
I am not really following what file is being overwritten

**Comment by heroin-moose** [comment]
I mean that built-in bash search function will be replaced with fzf. I'm not sure everyone want that.

**Comment by rafspiny** [comment]
@heroin-moose I would argue that replacing the built-in bash search function is the expected behaviour when installing fzf. I would expect to see fzf when I hit CTRL-R

Conflicts with other packages that overwrite the built-in bash search function are another matter

**Comment by heroin-moose** [comment]
I, for one, use it as vim plugin and not for shell autocompletion. So I would rather not.

**Comment by juippis** [comment]
This package is becoming such a nightmare already to handle... \***sigh**\*

Does it do the same for zsh and fish? I'm thinking at least bash-completion should be behind USE flag then, and the USE flag must describe this behaviour in metadata.xml.

**Comment by heroin-moose** [comment]
@juippis, I suggest following Arch way. Installing completions is okay. Keybindings, on the other hand, go to /usr/share/fzf where people can source them if they want.

**Comment by rafspiny** [comment]
I use it for both bash and vim. In my original commit, they were under a USE flag.
Debian is also adhering to what you said @heroin-moose 
```
Bash
====

Append this line to ~/.bashrc to enable fzf keybindings for Bash:

   source /usr/share/doc/fzf/examples/key-bindings.bash

Zsh
===

Append this line to ~/.zshrc to enable fzf keybindings for Zsh:

   source /usr/share/doc/fzf/examples/key-bindings.zsh

Fish
====

Issue the following commands to enable fzf keybindings for Fish:

   mkdir -p ~/.config/fish/functions/
   ln -s /usr/share/doc/fzf/examples/fish_user_key_bindings.fish ~/.config/fish/functions/

Vim
===

The most straightforward way is to append the following line to your vimrc:

   source /usr/share/doc/fzf/examples/fzf.vim

```
Shall I do the same and add an `einfo` print. I would do the same for fish and zsh now.

**Pull request closed without merging**